### DISCUSSION TITLE
Sending command with a whitespace at the end

### DISCUSSION THREAD

**Discussion by amitsova** [ktbyers/netmiko#2948] [Q&A]
Hi all,

Can somebody please help me? I am trying to send the "show mac add | i Gig1/0/1 ' command with white space at the end, because otherwise other interfaces are matched too (for example, Gig1/0/10, Gig1/0/11 and so forth and so on). I need only Gig1/0/1, but it looks like that the pattern that netmiko uses ignores the white space:

DEBUG:netmiko:read_channel: 
DEBUG:netmiko:write_channel: b'show mac add | i Gi1/0/1\n'
DEBUG:netmiko:Pattern is: show\ mac\ add\ \|\ i\ Gi1/0/1
DEBUG:netmiko:_read_channel_expect read_data: s
DEBUG:netmiko:_read_channel_expect read_data: how mac add | i Gi1/0/1

Here's my code:

```python
#!/usr/bin/env python3

from netmiko import ConnectHandler
import re
import logging
logging.basicConfig(filename='LOG.log', level=logging.DEBUG)
logger = logging.getLogger("netmiko")

with open('switches') as f:
    switches = f.read().splitlines()


for i in switches:
    dev_info = {'device_type': 'cisco_ios', 'ip': i, 'username': 'username', 'password': 'password'}
    net_connect = ConnectHandler(**dev_info)
    print(f'\nConnecting to {i}\n')
    show_int_status = net_connect.send_command('show int status | ex dis|not|%')
    ints_list = re.findall(r'^\w{2}\d\S+', show_int_status, re.M)
    ints_list = [f'{x} ' for x in ints_list]
    print(f'\n{show_int_status}\n\n{ints_list}\n')
    for j in ints_list:
        show_mac = net_connect.send_command(f"show mac add | i {j}      ")
        print(f'\n{len(show_mac)}\n\n{show_mac}')`
```

**Comment by ktbyers** [comment]
Netmiko's send_command automatically strips trailing whitespace by default.

Try:

```python
        show_mac = net_connect.send_command(f"show mac add | i {j}      ", normalize=False)
```

**Reply by amitsova** [reply]
Hi Kirk,

Thanks a lot! 

Unfortunately it fails with:


_Traceback (most recent call last):
  File "./collecting_static_macs1.1.py", line 24, in <module>
    show_mac = net_connect.send_command(f"show mac add | i {j}      ", normalize=False)
  File "/usr/local/lib/python3.6/site-packages/netmiko/utilities.py", line 363, in wrapper_decorator
    return func(self, *args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/netmiko/base_connection.py", line 1488, in send_command
    search_pattern
**OSError: Search pattern never detected in send_command_expect:** SLAN\-KAZ\-04\#_

I tried to set different `expect_string` couple of times, but it did not help. The output on the device itself  (of `"show mac add | i Gi1/0/1 "`) is empty, even though the port is connected:

SLAN-KAZ-04#show int status | i Gi1/0/1 
Gi1/0/1   DEV_DESCR_VMR1_ connected    26         a-full  a-100 10/100/1000BaseTX
SLAN-KAZ-04#
SLAN-KAZ-04#show mac add | i Gi1/0/1 
SLAN-KAZ-04#

For other connected interfaces, though, we can gather the mac address (having a command output), so I am not sure what would be the expect_string.
Do you have a clue? 

Thanks in advance!

p.s. I tried `net_connect.send_command_timing` - received no output when printing the `show` command. I tried `expect_string=r"#"` and after it `net_connect.set_base_prompt()`, no luck unfortunately.

**Reply by ktbyers** [reply]
Can you show me the log file when you use just this:

```python
        show_mac = net_connect.send_command(f"show mac add | i {j}      ", normalize=False)
```

**Reply by amitsova** [reply]
[LOG_1.log](https://github.com/ktbyers/netmiko/files/9579148/LOG_1.log)

Absolutely! Here it is.

**Reply by ktbyers** [reply]
Try adding a newline at the end:

```python
  show_mac = net_connect.send_command(f"show mac add | i {j}      \n", normalize=False)
```

Basically the `normalize=False` means Netmiko is no longer automatically adding that.

**Reply by amitsova** [reply]
[LOG_2.log](https://github.com/ktbyers/netmiko/files/9579227/LOG_2.log)
It behaves like `send_command_timing`. It works, however it does not print any outputs of the command for the interfaces.

Below's the code and the log is attached.

```python
#!/usr/bin/env python3

from netmiko import ConnectHandler
import re
import logging

logging.basicConfig(filename='LOG_2.log', level=logging.DEBUG)
logger = logging.getLogger("netmiko")

with open('switches') as f:
    switches = f.read().splitlines()

for i in switches:
    dev_info = {'device_type': 'cisco_ios', 'ip': i, 'username': 'username', 'password': 'password'}
    net_connect = ConnectHandler(**dev_info)
    print(f'\nConnecting to {i}\n')
    show_int_status = net_connect.send_command('show int status | ex dis|not|%')
    ints_list = re.findall(r'^\w{2}\d\S+', show_int_status, re.M)
    ints_list = [f'{x} ' for x in ints_list]
    print(f'\n{show_int_status}\n\n{ints_list}\n')
    for j in ints_list:
        # show_mac = net_connect.send_command(f"show mac add | i {j}")
        show_mac = net_connect.send_command(f"show mac add | i {j}      \n", normalize=False)
        # print(f'\n{len(show_mac)}\n')
        print(f'\n{show_mac}\n')
```

**Reply by ktbyers** [reply]
What does just `show mac add` by itself yield for this one switch?

You stated this above:

```
The output on the device itself (of "show mac add | i Gi1/0/1 ") is empty, even though the port is connected:
```

Can you expand on this a bit? I read this as you are not getting anything back on this when you do this manually using the CLI.

**Reply by amitsova** [reply]
Absolutely right. Here's a screenshot from the device itself, where the command is with whitespace at the end.
![Capture](https://user-images.githubusercontent.com/39456433/190566323-a04d1bbf-388e-47ea-b7c0-84693ed7118e.PNG)

The output of just `show mac add` shows the full mac table and it works ok. My purpose is to gather the "static" mac adds of end devices connected to the switch. It's a weird behavior to have the port connected and no mac shown :).
